orttraining/orttraining/test/graph/gradient_graph_builder_test.cc

```diff
@@ -1115,7 +1115,113 @@ void RetrieveSendRecvOperators(
   }
 }
 
-TEST(GradientGraphBuilderTest, PipelineOnlinePartition) {
+TEST(GradientGraphBuilderTest, PipelineOnlinePartition_bert_tiny) {
```

Comment:
Is it possible to have some smaller unit tests for this functionality, to ensure that we have full coverage for the code paths in the transformation? Off the top of my head, it would be good to see how we handle multiple shared initializers with different usage patterns, and also different usage patterns (e.g. stages that do all three of receiving, using, and sending a value). 

The end-to-end test here is nice to have, but it would be good to have smaller tests to confirm correctness as we modify the code.

Reply:
I agree. Currently bert is the only model with this scenario and we don't have smaller models we can test on. To create smaller unit tests for all possible permutations, I probably need to write a python script to generate those test models first. Let me work on this in another PR and I will create a task to track this too. 

---
In reply to: [449310150](https://github.com/microsoft/onnxruntime/pull/4321#discussion_r449310150) [](ancestors = 449310150)

Reply:
task added: [https://msdata.visualstudio.com/Vienna/_workitems/edit/809563](https://msdata.visualstudio.com/Vienna/_workitems/edit/809563)


---
In reply to: [451199963](https://github.com/microsoft/onnxruntime/pull/4321#discussion_r451199963) [](ancestors = 451199963,449310150)

Reply:
When looking at this in the future, how practical is it to generate models as part of the test case (just building the data structures in memory, as opposed to having Python generate them externally)?  If that approach works here it can help make the tests clearer by having the model directly next to the assertions being made about how it has been transformed.

Reply:
I am not sure how practical that would be... Currently only the operator test constructs small graphs (usually just contain one op) on the fly in test. I probably would need to do some research to figure it out. 

---
In reply to: [452230308](https://github.com/microsoft/onnxruntime/pull/4321#discussion_r452230308) [](ancestors = 452230308)